server/src/main/java/org/elasticsearch/action/bulk/TransportSimulateBulkAction.java

```diff
@@ -94,6 +109,52 @@ protected void doInternalExecute(
         );
     }
 
+    /**
+     * This creates a temporary with the mappings of the index in the request, and then attempts to index the source from the request
+     * into it. If there is a mapping exception, that exception is returned. On success the returned exception is null.
+     * @param request The IndexReqeust whose source will be validated against the mapping (if it exists) of its index
+     * @return a mapping exception if the source does not match the mappings, otherwise null
+     */
+    private Exception validateMappings(IndexRequest request) {
+        final SourceToParse sourceToParse = new SourceToParse(
+            request.id(),
+            request.source(),
+            request.getContentType(),
+            request.routing(),
+            request.getDynamicTemplates(),
+            DocumentSizeObserver.EMPTY_INSTANCE
+        );
+
+        ClusterState state = clusterService.state();
+        Exception mappingValidationException = null;
+        IndexAbstraction indexAbstraction = state.metadata().getIndicesLookup().get(request.index());
+        if (indexAbstraction != null) {
+            IndexMetadata imd = state.metadata().getIndexSafe(indexAbstraction.getWriteIndex(request, state.metadata()));
+            try {
+                indicesService.withTempIndexService(imd, indexService -> {
+                    indexService.mapperService().updateMapping(null, imd);
+                    return IndexShard.prepareIndex(
+                        indexService.mapperService(),
+                        sourceToParse,
+                        SequenceNumbers.UNASSIGNED_SEQ_NO,
+                        -1,
+                        -1,
+                        VersionType.INTERNAL,
+                        Engine.Operation.Origin.PRIMARY,
+                        Long.MIN_VALUE,
+                        false,
+                        request.ifSeqNo(),
+                        request.ifPrimaryTerm(),
+                        0
+                    );
+                });
+            } catch (Exception e) {
```

Comment:
Do we want to verify that it's a mapping exception? (Not sure what to do if it's not :man_shrugging: )

Reply:
Yeah I went back and forth on this. I thought about just logging a non-mapping exception. But then I thought a user would probably like to know about exceptions that are getting hit, even if they're not mapping exceptions. There might be something going wrong that isn't a mapping exception that a user can do something about. So I put it back this way.